### DISCUSSION TITLE
Sqlalchemy-2 discrepancy in return type for postgres ranges

### DISCUSSION THREAD

**Discussion by HawkOnPK** [sqlalchemy/sqlalchemy#11646] [Usage Questions] [answered]
I have a column of type `sqlalchemy.dialects.postgresql.DATERANGE`:
`validrange = Column(DATERANGE, nullable=False)`

When querying with a raw sql expression, i receive the expected type `psycopg2._range.DateRange` for the column:
`db.session.execute(text('SELECT validrange FROM table WHERE id = 1')).scalar()`
The same result is achieved by 
`db.session.execute(select(text('validrange')).where(Table.usageid == 1)).scalar()`

But when querying for the the declarative tables column, the return type changes to `sqlalchemy.dialects.postgresql.ranges.Range`, breaking various places:
`db.session.execute(select(Table.validrange).where(Table.usageid == 1)).scalar()`

In sqlalchemy 1.4 it did work to query for the `*Table.__table__.c` columns. Is this no longer supported in v2.0? What would be a declarative approach in this case now?

**Comment by lelit** [comment]
I'm not sure, but I think that the code path involving `text()` does not go thru the _adaptors_ that convert the driver specific datatype to the SA "generic" one.
But what do you mean with "breaking various places"?

**Reply by HawkOnPK** [reply]
The "breaking various places" was only referencing my own code, comparing sqlalchemy 1.4 and 2.0 behaviour.

Is there an option to get sqlalchemy to always return the psycopg object (or my own) instead of the generic "Range" one?

**Reply by CaselIT** [reply, marked as answer]
You would have to use your own type decorator that overrides the result_processor logic.
